packages/vitest/src/node/test-run.ts

```diff
@@ -0,0 +1,161 @@
+import type { File as RunnerTestFile, TaskResultPack } from '@vitest/runner'
+import type { SerializedError } from '../public/utils'
+import type { UserConsoleLog } from '../types/general'
+import type { Vitest } from './core'
+import type { TestProject } from './project'
+import type { HookOptions, TestCase, TestModule } from './reporters/reported-tasks'
+import type { TestSpecification } from './spec'
+
+export class TestRun {
+  private tests = emptyCounters()
+  private suites = emptyCounters()
+
+  constructor(private vitest: Vitest) {}
+
+  async start(specifications: TestSpecification[]) {
+    this.tests = emptyCounters()
+    this.suites = emptyCounters()
+    this.suites.total = specifications.length
+
+    await this.vitest.report('onTestRunStart', specifications)
+  }
+
+  async enqueued(project: TestProject, file: RunnerTestFile) {
+    this.vitest.state.collectFiles(project, [file])
+    const testModule = this.vitest.state.getReportedEntity(file) as TestModule
+    await this.vitest.report('onTestModuleQueued', testModule)
+  }
+
+  async collected(project: TestProject, files: RunnerTestFile[]) {
+    this.vitest.state.collectFiles(project, files)
+    await Promise.all([
+      this.vitest.report('onCollected', files),
+      ...files.map((file) => {
+        const testModule = this.vitest.state.getReportedEntity(file) as TestModule
+        return this.vitest.report('onTestModuleCollected', testModule)
+      }),
+    ])
+  }
+
+  async log(log: UserConsoleLog) {
+    this.vitest.state.updateUserLog(log)
+    await this.vitest.report('onUserConsoleLog', log)
+  }
+
+  async updated(update: TaskResultPack[]) {
+    this.vitest.state.updateTasks(update)
+
+    // These are used to guarantee correct reporting order
+    const runningTestModules: TestModule[] = []
+    const finishedTestModules: TestModule[] = []
+
+    const runningTestCases: TestCase[] = []
+    const finishedTestCases: TestCase[] = []
+
+    const startingHooks: HookOptions[] = []
+    const endingHooks: HookOptions[] = []
+
+    for (const [id,,,events] of update) {
+      const task = this.vitest.state.idMap.get(id)
+      const entity = task && this.vitest.state.getReportedEntity(task)
+
+      if (!entity) {
+        continue
+      }
+
+      for (const event of events) {
+        if (event === 'suite-prepare' && entity.type === 'module') {
+          runningTestModules.push(entity)
+        }
+
+        if (event === 'suite-finished' && entity.type === 'module') {
+          finishedTestModules.push(entity)
+
+          // Skipped tests need to be reported manually once test module has finished
+          for (const test of entity.children.allTests()) {
+            if (test.result().state === 'skipped') {
```

Comment:
do we need to report skipped tests? they won't have `start` and calling them arbitrary at the end seems weird to me

Reply:
We can report them when any `suite-finished` is emitted so it's at least timed with other tests, although the order is still wrong, - and they won't have `start` event. Also if suite skips all tests, `suite-finished` is never fired 🤔 

I don't know what to do here, honestly (a new hook?)

Reply:
I've now implemented this so that skipped test cases don't have any special rules. They are reported exactly as other test cases too. This will be easier for reporters to implement.

If this is not something we want, it can still be changed. 

https://github.com/vitest-dev/vitest/blob/41985f3d90f5d1de10233122f794ae2a1fbf8354/test/reporters/tests/test-run.test.ts#L24-L62

Reply:
IIRC the ui having some problems with todo tests since they are not reported in current api version, we need to change mode to `todo` on received end signal:

https://github.com/vitest-dev/vitest/blob/main/packages/ui/client/composables/explorer/collector.ts#L109

Maybe a bug in the ui where we set to running all tests when we click any run button.

Reply:
how can test trigger an event called "start" and "end" when it neither started nor ended?

Reply:
I green lit the names "start" and "end" when skipped tests weren't reported. the "start" event was called "prepare" because I assumed that skipped tests are reported and it doesn't make sense to call it start

We need to find a better name if we report them like that

Reply:
Would be nice to also have a test for the reverse:

```ts
       test.skip('skipped', () => {}); 
       test('running', () => {}); 
```

Reply:
> how can test trigger events called "start" and "end" when it neither started nor ended?

I guess we should re-check the idle logic in the client module, since we set the state to running when we receive on
TaskUpdated rpc callback.

We have onFinished, shouldn't be the end event?

Reply:
> how can test trigger events called "start" and "end" when it neither started nor ended?

In `test("skipped test", ctx => { ctx.skip(); })` for example. Though would it have "end" then?

Reply:
> Though would it have "end" then?

To clarify, `test("skipped test", ctx => { ctx.skip(); })` always has a start event, so not having an "end" wouldn't make sense

`test.skip()` doesn't have a start event, we just skip it

So, I think we should either:
- rename methods
- add different methods
- don't report skipped tests

Reply:
Any of those options sound good to me.

When building the summary reporter, I had difficulties showing correct test count when tests were skipped. The reporter keeps increasing test count when a test is reported as finished. But as skipped tests are not, I had to manually extract the count of skipped tests when test module finished. So the active test counter logic needs to be in multiple places.

Reply:
Maybe we need to have `queued/`dequeued` for tests like [Node has](https://nodejs.org/api/test.html#custom-reporters)? They will be called for every test, but `start`/`end` will only be called for unskipped tests

Reply:
(I think they need queued/dequeued because they run as the collect(?))

Reply:
Interestingly, `'test:start'` is emitted for tests _and_ subtests which is basically a suite?